### PR TITLE
Adds design document for an Omega error handler

### PR DESCRIPTION
**Pull request by philipwjones**
Design for an error handler that builds on the existing Log functionality. Comments and suggestions welcome.

Checklist
* [x] Documentation:
  * [x] Design document has been generated and added to the docs
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected

### CODE REVIEW

**Comment by cbegeman** [comment]
Looks reasonable to me. Thanks @philipwjones!

**Comment by vanroekel** [comment]
Looks great to me too, thanks @philipwjones!

**Comment by xylar** [comment]
Sorry for the delay.  I will review this later today.

**Review by mwarusz — COMMENTED** [review]
@philipwjones 
This design looks good to me, but I would like to have an even simpler interface for handling common situations like verifying preconditions. To give a concrete example, let's consider the following tracer code that checks the input time level:
https://github.com/E3SM-Project/Omega/blob/1e5e522de43cbac8fac832567f80e510738acfa4/components/omega/src/ocn/Tracers.cpp#L292-L296
Assuming that this error doesn't need to be propagated, ideally I would like to be able to write code like:
```
ERROR_REQUIRE(TimeLevel > 0 || (TimeLevel + NTimeLevels) <= 0, "Tracers: Time level {} is out of range", TimeLevel);
```
I think this could be easily build from the proposed macros.

Additionally, I am not sure about this, but I am wondering if we need to have a category of error checks that are disabled in release builds for performance reasons.

**Comment by xylar** [comment]
@mwarusz, your suggestion above looks like an assert if I understand correctly.  That seems like a good capability to include.  Would you have a failure result in a critical error?

**Comment by mwarusz** [comment]
@xylar 
Yes, this is meant to be some form of assert. However, in C++ asserts are typically disabled in Release mode, which motivated my second question. I think it should result in a critical error.

File: components/omega/doc/design/Error.md [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+```c++
+ERROR_SUM(ErrSum, ErrCode, ErrMsg, added vars if needed);
+```
+which after printing the messages, adds the most recent
+error code to the running sum stored in ``ErrSum``.
+
+Note that in the latter two cases, care must be taken so that
+the accumulated error or return codes do not match a
+predefined code and cause confustion in a later use of the
+code in an ``ERROR_CRITICAL`` or ``ERROR_CHECK`` call.
```
[the remark this answers is not in the document]
Comment (philipwjones) [reply components/omega/doc/design/Error.md:264]: Yeah, that's one approach. No approach is likely to absolutely guarantee it (eg a large predefined number added to a smaller error code may still hit another large predefined number), but we can do what we can to make it unlikely with some combination of gaps, allowed ranges.

**Comment by philipwjones** [comment]
@mwarusz  Thanks for the comments. Yes, I had considered an interface in which the condition was an argument so we didn't need the explicit conditional. There are some tricky bits about expanding something with an expression and still keeping the line number correct in the message/trace, but it can be done. I can add that to the doc.

**Review by xylar — APPROVED** [review]
Thanks for the clarifications, @philipwjones. That all sounds good.

**Comment by philipwjones** [comment]
@mwarusz I just pushed a change to add both a REQUIRE and ASSERT interface.  Feel free to suggest further changes or approve if this meets your needs.

**Review by mwarusz — APPROVED** [review]

**Pull request merged**